src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,146 @@
+            if (Objects.equals(key1, key2)) {
+                Object value1 = getValue();
+                Object value2 = entry.getValue();
+                return Objects.equals(value1, value2);
+            }
+            return false;
+        }
+
+        public final int hashCode() {
+            return (key == null ? 0 : Math.abs(key.hashCode()));
+        }
+    }
+
+    private int indexFor(K key) {
+        return key == null ? 0 : Math.abs(key.hashCode() % DEFAULT_CAPACITY);
+    }
+
+    private void putForNullKey(V value) {
+        for (Entry<K, V> entry = table[0]; entry != null; entry = entry.next) {
+            if (entry.key == null) {
+                entry.value = value;
+                return;
+            }
+        }
+        addEntry(null, value, 0);
+    }
+
+    private void addEntry(K key, V value, int index) {
+        Entry<K, V> entry = table[index];
+        table[index] = new Entry<>(key, value, entry);
+        if (size++ >= threshold) {
+            resize(2 * table.length);
+        }
+    }
+
+    private void resize(int newCapacity) {
+        Entry<K, V>[] newTable = new Entry[newCapacity];
+        transfer(newTable);
+        table = newTable;
+        threshold = (int) (newCapacity * LOAD_FACTOR);
+    }
+
+    private void transfer(Entry<K, V>[] newTable) {
+        Entry<K, V>[] src = table;
+        for (int j = 0; j < src.length; j++) {
+            Entry<K, V> entry = src[j];
+            if (entry != null) {
+                src[j] = null;
+            }
+            while (entry != null) {
+                Entry<K, V> next = entry.next;
+                int i = indexFor(entry.key);
+                entry.next = newTable[i];
+                newTable[i] = entry;
+                entry = next;
+            }
+        }
+    }
+
+    private V getForNullKey() {
+        for (Entry<K, V> entry = table[0]; entry != null; entry = entry.next) {
+            if (entry.key == null) {
+                return entry.value;
+            }
+        }
+        return null;
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (key == null) {
+            putForNullKey(value);
+        } else {
+            int index = indexFor(key);
+            for (Entry<K, V> entry = table[index]; entry != null; entry = entry.next) {
+                if (key == entry.key || key.equals(entry.key)) {
+                    entry.value = value;
+                    return;
+                }
+            }
+            addEntry(key, value, index);
+        }
     }
 
     @Override
     public V getValue(K key) {
-        return null;
+        if (key == null) {
+            return getForNullKey();
+        } else {
+            for (Entry<K, V> entry = table[indexFor(key)]; entry != null; entry = entry.next) {
+                if (key == entry.key || key.equals(entry.key)) {
+                    return entry.value;
+                }
+            }
+            return null;
+        }
```

Comment:
you don't need separate method for null key, use only one for loop for getting value by key